File: npm/tests/codex.test.js
'use strict'

const test = require('node:test')
const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')
const { installCodex } = require('../src/generators/codex')

test('installCodex creates idempotent native Codex artifacts', () => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-codex-'))
  installCodex(root)
  installCodex(root)

  const required = [
    'AGENTS.md',
    '.codex/config.toml',
    '.codex/hooks.json',
    '.codex/agents/trackfw-architect.toml',
    '.codex/agents/trackfw-reviewer.toml',
    '.agents/skills/trackfw-governance/SKILL.md',
    '.agents/skills/trackfw-release/SKILL.md',
  ]
  for (const name of required) assert.equal(fs.existsSync(path.join(root, name)), true, name)

  const config = fs.readFileSync(path.join(root, '.codex/config.toml'), 'utf8')
  assert.equal((config.match(/^\[agents\]$/gm) || []).length, 1)
  assert.match(config, /max_threads = 6/)
  assert.match(config, /max_depth = 1/)

  const hooks = JSON.parse(fs.readFileSync(path.join(root, '.codex/hooks.json'), 'utf8')).hooks
  assert.equal(hooks.PermissionRequest.length, 1)
  assert.equal(hooks.PermissionRequest[0].hooks[0].command, 'scripts/trackfw-attention-signal.sh')
  assert.equal(hooks.PostToolUse.length, 1)
  assert.equal(hooks.PostToolUse[0].hooks[0].command, 'scripts/trackfw-attention-cleanup.sh')
  assert.equal(hooks.PreToolUse, undefined)
})

File: internal/generators/codex_test.go
package generators

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestInstallCodexCreatesNativeArtifacts(t *testing.T) {
	dir := t.TempDir()

	if err := InstallCodex(dir); err != nil {
		t.Fatalf("InstallCodex() error: %v", err)
	}
	if err := InstallCodex(dir); err != nil {
		t.Fatalf("InstallCodex() must be idempotent: %v", err)
	}

	required := []string{
		"AGENTS.md",
		".codex/config.toml",
		".codex/hooks.json",
		".codex/agents/trackfw-architect.toml",
		".codex/agents/trackfw-backend.toml",
		".codex/agents/trackfw-frontend.toml",
		".codex/agents/trackfw-qa.toml",
		".codex/agents/trackfw-security.toml",
		".codex/agents/trackfw-reviewer.toml",
		".agents/skills/trackfw-governance/SKILL.md",
		".agents/skills/trackfw-plan/SKILL.md",
		".agents/skills/trackfw-implement/SKILL.md",
		".agents/skills/trackfw-review/SKILL.md",
		".agents/skills/trackfw-release/SKILL.md",
	}
	for _, name := range required {
		if _, err := os.Stat(filepath.Join(dir, name)); err != nil {
			t.Errorf("missing Codex artifact %s: %v", name, err)
		}
	}

	config, err := os.ReadFile(filepath.Join(dir, ".codex", "config.toml"))
	if err != nil {
		t.Fatal(err)
	}
	if strings.Count(string(config), "[agents]") != 1 ||
		!strings.Contains(string(config), "max_threads = 6") ||
		!strings.Contains(string(config), "max_depth = 1") {
		t.Errorf("unexpected Codex config:\n%s", config)
	}

	hooks := readJSONFile(t, filepath.Join(dir, ".codex", "hooks.json"))
	if !hasClaudeHookEntry(hooks, "PermissionRequest", ".*", "scripts/trackfw-attention-signal.sh") {
		t.Error("PermissionRequest attention hook not found")
	}
	if !hasClaudeHookEntry(hooks, "PostToolUse", ".*", "scripts/trackfw-attention-cleanup.sh") {
		t.Error("PostToolUse cleanup hook not found")
	}
	if hasClaudeHookEntry(hooks, "PreToolUse", ".*", "scripts/trackfw-attention-signal.sh") {
		t.Error("Codex attention must use PermissionRequest, not PreToolUse")
	}
}
